### ISSUE TITLE
[2.5.x] Multipart Form Upload with empty body will create empty file

### ISSUE THREAD

**Issue by schmitch** [playframework/playframework#6203]
### Play Version (2.5.x / etc)

2.5.x not tested others
### API (Scala / Java / Neither / Both)

Tested Scala but I guess Java has this issue aswell.
### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

Mac OS X 10.11.5
### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

java version "1.8.0_92"
Java(TM) SE Runtime Environment (build 1.8.0_92-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.92-b14, mixed mode)
### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.
1. Submit a form with a empty file
2. Browser will still add a Multipart Part for it:

```
------WebKitFormBoundarySVvXYhLnZ0OUY2jV
Content-Disposition: form-data; name="image"; filename=""
Content-Type: application/octet-stream


------WebKitFormBoundarySVvXYhLnZ0OUY2jV--
```
1. Excpect that request.body.file("partname") will be None
### Actual Behavior

Actually a browser will submit a empty body when you submit a multipart form. Play will parse that as a empty file.
Actually this behavior is unexpected since the body part is empty and that means "nothing" should be created, returning a empty file is definitiv not a good idea here.
The problem is that we don't know if more data is coming here but I guess we should check if the body is empty (i.e. the body part will contain a CrLf).
WDYT?

Edit: Maybe we could add such cases to badParts.

**Comment by mkurz** [comment]
I agree we shouldn't create an empty file in this case.

**Comment by rania-globalorange** [comment]
I can confirm about Java, I ran into this too.

**Comment by wsargent** [comment]
Could you work around this with a custom file part handler?  https://github.com/playframework/play-fileupload-scala

**Comment by nicolov** [comment]
Hey there, I've run across this as well. Is checking for a non-empty file name enough to work around this? (I'm thinking about a single file field).

**Comment by rania-globalorange** [comment]
@nicolov Yes, this is what I did as a quick fix

**Comment by JFCote** [comment]
This is a terrible regression from 2.4 to 2.5. Now I need to add validation everywhere.
When is this going to be fixed?

**Comment by schmitch** [comment]
> This is a terrible regression from 2.4 to 2.5. Now I need to add validation everywhere.
> When is this going to be fixed?

It would be sad if you didn't had validation **before**. It's not like file upload is one of the most critical things when it comes to security.

**Comment by JFCote** [comment]
We have validations. We check if the FilePart is different than null (plus other validation after that). But now, a lot of stuff that were using the getFilename() after the initial validation started to crash because the FilePart is not null and the filename is empty. There is also a generated temporary file on disk of 0 byte. So everywhere where I checked that the FilePart was null, I added a check if the length of the generated temporary file is different than 0.

To me, it's not logic that there is a "generated" file on disk of 0 byte since there is no file at all in the original form! The FilePart should be equal to null like in 2.4 and before.

This should at least be added to the documentation of migration from 2.4 to 2.5 :)

**Comment by schmitch** [comment]
well it's a bug (after all), but as said it's not a critical one, since you actually **should** check the file for various things anyway. Even without the bug somebody could generate zero-one byte files without a filename ;)
Maybe that's a different beast that we actually, don't talk about security in our file upload chapters ;)

**Comment by shkhln** [comment]
> It's not like file upload is one of the most critical things when it comes to security.

*is not* ftfy

> lot of stuff that were using the getFilename() after the initial validation started to crash because the FilePart is not null and the filename is empty

What are you doing with the filename? There is probably a genuine security issue in you application.

**Comment by JFCote** [comment]
We are not talking about security issue in my application. Maybe my application is a small internal app used by 4 people who will not try to send a 0 byte file with an empty name to break it, maybe not.

Your comments make me feel like you don't care about fixing this bug, and that's ok, you are the contributor to this open source project after all. I have nothing to say about this.

But your passive-agressive attitude of turning the problem back on me is unacceptable in the open source world. I take my time to go online, search for a bug report, add my experience on it and instead of telling me that you are going to check it someday, you are telling me that I have security issue in my app.

From what I know of software development, if between 2 versions, there is an unexpected behavior that break code and that is not documented in the migration documentation, it's called a bug.

Anyway, happy holidays

**Comment by shkhln** [comment]
> We are not talking about security issue in my application. Maybe my application is a small internal app used by 4 people who will not try to send a 0 byte file with an empty name to break it, maybe not.

You are correct that it has nothing to do with this bug. However, "crash" sounds very suspicious and I was just thinking I could likely suggest a proper fix (to the crash) if I know more details. For what it's worth, I actually thought that simply pointing you to https://www.owasp.org/index.php/Unrestricted_File_Upload might be considered too rude.

**Comment by zhangway** [comment]
I've fixed it and tested it.

**Comment by mkurz** [comment]
@zhangway What do you mean? Did you patch the framework?

**Comment by zhangway** [comment]
@mkurz No. I just used the simple check `if filename.isEmpty`.

**Comment by kheraud** [comment]
Just having the same issue with 2.5.12. I chose a workaround with `filename.isEmpty`

**Comment by mkurz** [comment]
#5037 introduced this bug: Before that pull request a `MultipartFormData` also collected files with empty filename in a `missingFileParts: Seq[MissingFilePart])`:
https://github.com/playframework/playframework/pull/5037/files#diff-a41a1796efe45be8ecc6d68806de1b81L151
That `missingFileParts` was removed, instead also those "missing file parts" will just be added like a "regular" file to the files sequence.

**Comment by mkurz** [comment]
Fixed via #8880 which checks if the filename is empty (which should be sufficient to fix that regression, because that was the check that was done in 2.4 as well to prevent this bug).
